### ISSUE TITLE
Stryker v4 Instrumentor breaks enzyme snapshots

### ISSUE THREAD

**Issue by Garethp** [stryker-mutator/stryker#2362]
**Summary**
When using Stryker v4, my snapshots (using enzyme) break because where in the main run it recognizes the component name, in the mutant run it just calls it `Component`.

Example: 

```tsx
const ErrorMessage = ({ message, id }: Props) => (
  <React.Fragment>
    {message.length > 0 && (
      <span id={id} className="error-message">
        {message}
        <span className="hidden">Error: {message}</span>
      </span>
    )}
  </React.Fragment>
);

export default ErrorMessage;
```

```tsx
interface IFormLabelProps {
  label: string;
  forId: string;
  errorMessage?: string;
}

const FormLabel: FunctionComponent<IFormLabelProps> = ({
  label,
  forId,
  errorMessage
}) => {
  return (
    <>
      <label className="label" htmlFor={forId}>
        {label}
      </label>
      {!!errorMessage?.length && (
        <ErrorMessage id={`${forId}-error`} message={errorMessage} />
      )}
    </>
  );
};
```

```tsx
it("should render with error message", () => {
  const formLabel = shallow(
    <FormLabel
      label="test label"
      forId="testId"
      errorMessage="test error message"
    />
  );
  expect(formLabel).toMatchSnapshot();
});
```

In the main run it will return the snapshot 

```js
exports[`should render with error message 1`] = `
<Fragment>
  <label
    className="label"
    htmlFor="testId"
  >
    test label
  </label>
  <ErrorMessage
    id="testId-error"
    message="test error message"
  />
</Fragment>
`;
```

While in the mutant it'll render

```js
exports[`should render with error message 1`] = `
<Fragment>
  <label
    className="label"
    htmlFor="testId"
  >
    test label
  </label>
  <Component
    id="testId-error"
    message="test error message"
  />
</Fragment>
`;
```

The difference being that `ErrorMessage` has been replaced with `Component`. This is because the instrumentor replaced the main definition of `ErrorMessage` with a ternary, and it appears that enzyme (or react, not sure) is looking for the variable name of the original variable that the component is assigned to. 

If you replaced ErrorMessage with

```tsx
const ErrorMessage = true ? ({ message, id }: Props) => (
  <React.Fragment>
    {message.length > 0 && (
      <span id={id} className="error-message">
        {message}
        <span className="hidden">Error: {message}</span>
      </span>
    )}
  </React.Fragment>
) : () => undefined;
```

Then the snapshot would also fail in the same way it does during mutation tests.

One possible fix is to temporarily assign the component to a variable with the original name inside the ternary. That way the component maintains its original variable name. Here's an example below:

```tsx
let ErrorMessage;

const ErrorMessageInstrumented = true
  ? (ErrorMessage = ({ message, id }: Props) => (
      <React.Fragment>
        {message.length > 0 && (
          <span id={id} className="error-message">
            {message}
            <span className="hidden">Error: {message}</span>
          </span>
        )}
      </React.Fragment>
    ))
  : () => undefined;

export default ErrorMessageInstrumented;
```

**Comment by nicojs** [comment]
Hi @Garethp thanks a lot for opening this issue. Are you still planning to add a e2e test with enzyme snapshots? It's optional of course, but it would help us a lot with solving this issue. It's also fine if you just add the files a new `e2e/test` directory (for example `e2e/test/jest-with-enzyme-snapshots`). If you need any help, don't hesitate to react here, or hit us up on slack

**Comment by Garethp** [comment]
Yes I am, I just didn't find the time this weekend. I should have a PR in the next day or so

**Comment by nicojs** [comment]
The problem is a little bit more broad even, as you can see here:

```
$ node -p "const Foo = class { }; Foo.name"
Foo
$ node -p "const Foo = true? class { }: ''; Foo.name"

```

**Comment by nicojs** [comment]
I've been thinking about this and I think I know of a solution that works for most use cases.

I want to instrument these:

```ts
const foo = function () { }
const bar = () => {}
const Baz = class { }
```

Like this:

```ts
const foo = true? function foo () { }: ...;
const bar = true? (() => { const bar = () =>{}; return bar } {})() : ...;
const Baz = true? class Baz { } : ...;
```

What do you think of this?

I don't like your approach @Garethp, because it changes the functionality of the code too much and will be difficult to maintain.

**Comment by Garethp** [comment]
That looks like it works quite  well. I'll test it on my work project when it's been pushed up

**Comment by nicojs** [comment]
I've just released `4.0.0-beta.3` with this fix. Feel free to give it another go @Garethp 🌹

**Comment by Garethp** [comment]
This is strange, but there are some cases where you might *not* want to be doing this? Take the following case:

```tsx
export const SomeItem = () => <div></div>;
```

For some reason in non-mutated code this gets exported *without* a name. In the mutated code it now *does* have a name? I didn't even notice this before.

Also, it looks like there's a bug where having a string that starts with `https://` as a return seems to break the instrumentor and it spits out an unterminated string. I'll try to get that as a e2e test in the next couple of days
